### PR TITLE
Circleci setup & copyright notice check

### LINKED ISSUE #49: Setup continous integration pipeline
<!-- Provide a general summary of the task in the title above.

When adding program output or code snippets in the issue, use code blocks with
appropriate syntax highlighting. -->

#### Motivation / Context
<!-- Tell us why this task is important. -->
To setup a continuous integration pipeline.

#### Description                                                                                                                                                  
<!-- Describe the task briefly. -->                                                                                                                               
The CI jobs can be setup on a [circleci](https://circleci.com/) platform. It has a free tier for open source projects. 

- [ ] Configure the pipeline for three jobs:                                                                                             
~1. Checkout.~ Fetching the dependencies directly during lint / test job results in a faster pipeline than using a dedicated container for tests. This is mainly due to the time taken (~15 to 20s) to spin a a dedicated container to fetch dependencies.
2. Run linter.                                                                                                                                                    
3. Run tests.                                                                                                                                                     

- [ ] In the repository, configure the jobs to be  triggered on every push operation on any branch.

- [ ] Display a badge in Readme.md file for CI pipeline status of `master` and `develop` branches.                                                                             
                                                                                                                                                                  
##### Relates to                                                                                                                                                  
<!-- Add links to any related issue or other references here. -->                                                                                                 
Depends on #48.
                                                                                                                                                                  
#### Testing                                                                                                                                                      
<!-- Describe what should be tested and how it can be done. -->                                                                                                   
-NA-
                                                                                                                                                                  
#### Thread Safety                                                                                                                                                
<!-- Should any component of the implementation be thread safe? If so, why? -->                                                                                   
-NA-

### PR DESCRIPTION
**Pull request by manoranjith**
#### Description

The script can be invoked from the project root using the command:
 ```
.scripts/copyright_notice_check.sh
 ```
It checks if the copyright header in each go file (excluding the generated mocks in "internal/mocks") match the template in the
"copyright_notice_template.txt" in the scripts dir. When there is a mismatch, it prints the diff.
    
~When a build tag is used in the go file, it detects it and looks for the copyright notice from line 3.~ Not required, as the copyright header can precede the build tag definition.

##### Category

Implementation Task

##### Relevant issue

Resolves #49 

##### Steps to run the tests

1. Correctness of copyright notice check can be tested by adding and file with wrong header and running the script.
2. Correctness of circle ci config can be validated using the circle ci local setup tool. Steps can be found [here](https://support.circleci.com/hc/en-us/articles/360006735753-Validating-your-CircleCI-Configuration).

#### Checklist 

- [x ] Name is added to the NOTICE file, if it is not present already.
- [x ] Changes are rebased onto the target branch.

### CODE REVIEW

**Comment by choeppler** [comment]
Hi @manoranjith, there's some conflict with this PR after #52 got merged. Could you rebase, please?

**Comment by manoranjith** [comment]
Hi @choeppler, I have now rebased the branch to latest develop.

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+jobs:
+
+  lint:
+    executor: dst-go-executor
+    steps:
+      - checkout-code
+      - make-results-artifacts-dir
+      - run:
+          name: Install linter
+          command: curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh| sh -s -- -b $(go env GOPATH)/bin v1.28.2
```
Comment (choeppler) [inline .circleci/config.yml:47]:
```suggestion
          command: curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.28.2
```
nitpick... consistent formatting

File: .scripts/copyright_notice_check.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
```
Comment (choeppler) [inline .scripts/copyright_notice_check.sh:1]: `check_copyright_notice()` would set `exit_status=1` for this file... ;-)

File: .circleci/config.yml (hunk 1 above)
Comment (manoranjith) [reply .circleci/config.yml]: Updated in e2db49f07a920b374190486bb35716ee39bc71b4.

File: .scripts/copyright_notice_check.sh (hunk 2 above)
Comment (manoranjith) [reply .scripts/copyright_notice_check.sh:1]: Updated in fdf95f0a6c9c2da61a6e9a16ddf3124409ddb7e5.

Comment (choeppler) [reply .scripts/copyright_notice_check.sh:1]: fixed in 0e2b98aa

**Pull request merged**